cluster-autoscaler/proposals/cp_agnostic_ca.md

```diff
@@ -0,0 +1,158 @@
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/digitalocean"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/external"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/gce"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/magnum"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/packet"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+)
+
+// AvailableCloudProviders supported by the cloud provider builder.
+var AvailableCloudProviders = []string{
+	cloudprovider.AwsProviderName,
+	cloudprovider.AzureProviderName,
+	cloudprovider.GceProviderName,
+	cloudprovider.AlicloudProviderName,
+	cloudprovider.BaiducloudProviderName,
+	cloudprovider.MagnumProviderName,
+	cloudprovider.DigitalOceanProviderName,
+	cloudprovider.ExternalCloudProviderName,
+	clusterapi.ProviderName,
+}
+
+// DefaultCloudProvider is GCE.
+const DefaultCloudProvider = cloudprovider.GceProviderName
+
+func buildCloudProvider(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	switch opts.CloudProviderName {
+	case cloudprovider.GceProviderName:
+		return gce.BuildGCE(opts, do, rl)
+	case cloudprovider.AwsProviderName:
+		return aws.BuildAWS(opts, do, rl)
+	case cloudprovider.AzureProviderName:
+		return azure.BuildAzure(opts, do, rl)
+	case cloudprovider.AlicloudProviderName:
+		return alicloud.BuildAlicloud(opts, do, rl)
+	case cloudprovider.BaiducloudProviderName:
+		return baiducloud.BuildBaiducloud(opts, do, rl)
+	case cloudprovider.DigitalOceanProviderName:
+		return digitalocean.BuildDigitalOcean(opts, do, rl)
+	case cloudprovider.MagnumProviderName:
+		return magnum.BuildMagnum(opts, do, rl)
+	case packet.ProviderName:
+		return packet.BuildPacket(opts, do, rl)
+	case clusterapi.ProviderName:
+		return clusterapi.BuildClusterAPI(opts, do, rl)
+	case cloudprovider.ExternalCloudProviderName:
+		return external.BuildExternalCloudProvider(opts, do, rl)
+	}
+	return nil
+}
+```
+
+- Introduce builder_external.go
+```go
+// +build external
+
+package builder
+
+import (
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/external"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+)
+
+// AvailableCloudProviders supported by the cloud provider builder.
+var AvailableCloudProviders = []string{
+	cloudprovider.ExternalCloudProviderName,
+}
+
+const DefaultCloudProvider = cloudprovider.ExternalCloudProviderName
+
+func buildCloudProvider(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+		return external.BuildExternalCloudProvider(ops,do,rl)
+}
+```
+
+- External cloudprovider will register their implementation using init block. (in their own repo)
+```go
+package ecp
+//ecp: ExternalCloudProvider
+import "k8s.io/autoscaler/cluster-autoscaler/cloudprovider/external"
+
+func init() {
```

Comment:
Could you detail how this new custom-autoscaler would bootstrap ? Perhaps the tests should cover this scenario creating an example of an external cloud provider.

Reply:
We will be able to add only test for the registration and wiring flow for external_cp. Actual execution of external_cp should be outside the purview of CA. More inputs are invited on this point.

Reply:
As a developer I suggest you add a sample-external-provider for two reasons:
* Test this new CA feature.
* Show to the community how to code a sample-external-provider. You can follow as example what was done in https://github.com/kubernetes/sample-controller.

Reply:
- Agreed to that point of writing a sample dummy external_cp. A sample and readme on ho to add will be helpful.

On testing 2 test can be included:
- failure scenario in bootup when external_cp is not registered
- success scenario when the dummy_external_cp is registered